## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(["EIN", "NAME", "USE_CASE",],1)

In [3]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
app_values = application_df["APPLICATION_TYPE"].value_counts()
app_values

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T14        3
T25        3
T15        2
T29        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(app_values[app_values < 950].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
Other     2266
T4        1542
T6        1216
T5        1173
T19       1065
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts for binning
app_class = application_df["CLASSIFICATION"].value_counts()
app_class

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C2600        1
C2170        1
C5200        1
C1236        1
C2570        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
app_class[app_class > 1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C6000       15
C1800       15
C1250       14
C8200       11
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C2400        6
C4100        6
C1720        6
C1257        5
C1600        5
C0           3
C1260        3
C2710        3
C1267        2
C1256        2
C3200        2
C1234        2
C1246        2
Name: CLASSIFICATION, dtype: int64

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list(app_class[app_class < 950].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
features = pd.get_dummies(application_df)

In [10]:
# Split our preprocessed data into our features and target arrays
X = features.drop("IS_SUCCESSFUL", axis="columns")
y = features["IS_SUCCESSFUL"]

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [12]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#attempted 100
#attempted 250
#added 3rd layer, attempting 375
#attempting 450
#attempting 550

hidden_layer1 = 550
hidden_layer2 = 550
hidden_layer3 = 550
no_features = len(X_train.columns)

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_layer1, input_dim=no_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_layer2, activation="relu"))

nn.add(tf.keras.layers.Dense(units=hidden_layer3, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 550)               19800     
_________________________________________________________________
dense_1 (Dense)              (None, 550)               303050    
_________________________________________________________________
dense_2 (Dense)              (None, 550)               303050    
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 551       
Total params: 626,451
Trainable params: 626,451
Non-trainable params: 0
_________________________________________________________________


In [13]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [14]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=75)

Epoch 1/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5758 - accuracy: 0.7184
Epoch 2/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5629 - accuracy: 0.7223
Epoch 3/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5615 - accuracy: 0.7243
Epoch 4/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5598 - accuracy: 0.7250
Epoch 5/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5592 - accuracy: 0.7267
Epoch 6/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5580 - accuracy: 0.7276
Epoch 7/75
804/804 [==============================] - 6s 8ms/step - loss: 0.5581 - accuracy: 0.7270
Epoch 8/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5568 - accuracy: 0.7283
Epoch 9/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5569 - accuracy: 0.7276
Epoch 10/75
804/804 [==============================] - 6s 7ms/step - loss: 0.5561 - accuracy: 0.7269

In [ ]:
#Attempt:
#-2 hidden layers @ 100, 100 epochs
#-2 hidden layers @ 250, 100 epochs (computer did not appreciate)
#-3 hidden layers @ 375, 60 epochs
#-3 hidden layers @ 450, 75 epochs
#-3 hidden layers @ 550, 75 epochs

In [15]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5994 - accuracy: 0.7250
Loss: 0.5994295477867126, Accuracy: 0.7250145673751831


In [16]:
# Export our model to HDF5 file
nn.save("AlphabetSoupCharity.h5")

I was not able to achieve a greater accuracy than 73% even after multiple adjustments to hidden node layers, removing another column, and changing the number of epochs.  I did not feel that I had enough understanding to adjust further variables in this data without removing something important.  The purpose of this notebook was to predict whether applicants to a charity, Alphabet Soup, would be successfully funded, however unfortunately we were not able to reach the required 75% accuraccy.